File: src/main/scala/at/pria/koza/polybuf/PolybufConfig.scala
/**
 * PolybufConfig.scala
 *
 * Created on 18.05.2013
 */

package at.pria.koza.polybuf

import scala.collection.mutable

import com.google.protobuf.ExtensionRegistry

/**
 * <p>
 * A `PolybufConfig` encapsulates `PolybufIO`s for different object types, each identified by an `Int` that
 * corresponds to the protobuf extension field label associated with that type (see `PolybufSerializable`). Note
 * that this class does not have a notion of what message type a `PolybufIO` corresponds to, and thus does not
 * enforce anything of the sort. Implementors are free to design `PolybufIO` classes that work for any number and
 * any type of messages. The user simply has to register that `PolybufIO` with all the messages they want.
 * </p>
 *
 * @version V1.0 18.05.2013
 * @author SillyFreak
 */
class PolybufConfig() {
  private val config = mutable.Map[Int, PolybufIO[_ <: PolybufSerializable]]()
  val registry: ExtensionRegistry = ExtensionRegistry.newInstance()

  def add(io: PolybufIO[_ <: PolybufSerializable]): Unit = {
    config.put(io.typeId, io)
    registry.add(io.extension)
  }

  def get(typeID: Int): Option[PolybufIO[_ <: PolybufSerializable]] =
    config.get(typeID)
}

File: src/main/scala/at/pria/koza/polybuf/PolybufConfig.java
/**
 * PolybufConfig.java
 * 
 * Created on 18.05.2013
 */

package at.pria.koza.polybuf;


import java.util.HashMap;
import java.util.Map;

import com.google.protobuf.ExtensionRegistry;


/**
 * <p>
 * A {@code PolybufConfig} encapsulates {@link PolybufIO}s for different object types, each identified by an
 * {@code int} that corresponds to the protobuf extension field label associated with that type (see
 * {@link PolybufSerializable}). Note that this class does not have a notion of what message type a
 * {@code PolybufIO} corresponds to, and thus does not enforce anything of the sort. Implementors are free to
 * design {@code PolybufIO} classes that work for any number and any type of messages. The user simply has to
 * register that {@code PolybufIO} with all the messages he wants.
 * </p>
 * 
 * @version V1.0 18.05.2013
 * @author SillyFreak
 */
public class PolybufConfig {
    private final Map<Integer, PolybufIO<?>> config;
    private final ExtensionRegistry          registry;
    
    public PolybufConfig() {
        config = new HashMap<>();
        registry = ExtensionRegistry.newInstance();
    }
    
    public ExtensionRegistry getRegistry() {
        return registry;
    }
    
    public void add(PolybufIO<?> io) {
        config.put(io.getType(), io);
        registry.add(io.getExtension());
    }
    
    @SuppressWarnings("unchecked")
    public <T> PolybufIO<T> get(int type) {
        return (PolybufIO<T>) config.get(type);
    }
}